### PR TITLE
add "melange for x devs" section

### LINKED ISSUE #11: Docs should explain `rescript-syntax`

### LINKED ISSUE #20: Add section "Melange for X users"
For ReScript users, for OCaml users, for TypeScript/JavaScript users.

### PR DESCRIPTION
**Pull request by jchavarri**
Fixes #20.
Fixes #11.

### CODE REVIEW

File: docs/melange-for-x-developers.md [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+```
+
+In OCaml, nominal typing is fully supported. Some of the core language types
+like [records and
+variants](https://v2.ocaml.org/manual/coreexamples.html#s%3Atut-recvariants) are
+nominal. This means that even if you declare exactly the same type twice,
+functions that operate on values from one type will not be compatible with the
+other type.
+
+There is also structural typing, used for OCaml
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:298]: and modules.

File: docs/melange-for-x-developers.md [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+### Strictness and soundness
+
+TypeScript you can use types like `any`, or other extensive types like
+`Function`. This can be mitigated using the `strict` option in the
+`tsconfig.json`. There is no such option in OCaml.
+
+TypeScript may sacrifice soundness for practicality when needed, like mentioned
+in [the
+handbook](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#a-note-on-soundness).
+OCaml implementations provide methods like `Obj.magic` but they are more
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:340]: and the identity primitive, which we also documented.

File: docs/melange-for-x-developers.md [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+  which is not possible with Melange.
+- `Marshal` is well supported in Js_of_ocaml, while Melange does not support it.
+- Libraries like `Unix` or `Str` are available in Js_of_ocaml but not in
+  Melange.
+- Js_of_ocaml supports sourcemaps, which Melange do not support yet (as of May
+  2023).
+
+On the upside, in Melange:
+
+- Consuming existing JavaScript packages might a bit easier in Melange, thanks
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:499]:
```suggestion
- Consuming existing JavaScript packages might be a bit easier in Melange, thanks
```

File: docs/melange-for-x-developers.md [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+  to its compilation model and the extensive availability of mechanisms to bind
+  to JavaScript code.
+- There is great support for some of the most used JavaScript libraries like
+  [ReactJS](https://github.com/reasonml/reason-react/) or GraphQL clients.
+- The generated JavaScript bundles are generally smaller.
+- The generated JavaScript code is generally more readable.
+- Straight-forward integration with JavaScript tooling like Webpack, NextJS,
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:506]:
```suggestion
- Straight-forward integration with modern JavaScript tooling like Webpack, NextJS,
```

File: docs/melange-for-x-developers.md [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+  etc. This is possible thanks to the 1 module <-> 1 JavaScript file compilation
+  model.
+
+## For ReScript developers
+
+As a project that was born originally from ReScript, Melange inherits a lot of
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:512]:
```suggestion
As a project that shares a common ancestry with ReScript, Melange inherits a lot of
```

File: docs/melange-for-x-developers.md [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+
+ReScript projects rely exclusively on npm for all packages they depend on.
+Melange projects, on the other hand, will use opam for native packages, and npm
+for JavaScript ones. Melange package management is explained in detail in [the
+dedicated section](package-management.md).
+
+### Build system
+
+While ReScript has its own build system, originally based on Ninja, Melange
+defers to Dune as orchestrator of builds, as is explained in detail in [the
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:535]:
```suggestion
defers to Dune for build orchestration, as is explained in detail in [the
```

File: docs/melange-for-x-developers.md [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+### Source-based vs pre-built distribution
+
+While with ReScript every dependency can be downloaded with just npm, Melange
+projects will have to use opam and npm. This is a trade-off: on one hand, some
+Melange projects might need to include two package configuration files. But on
+the other hand they can benefit from opam’s source-based package distribution
+model for things like PPXs, linters, or any other OCaml tooling.
+
+By comparison, to consume any OCaml ecosystem tool in ReScript is more
+challenging. As npm is a package manager designed for an interpreted language,
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:557]: "there's no native toolchain available in rescript" should be mentioned somewhere. That's the actual limitation

File: docs/melange-for-x-developers.md [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+maintenance harder for library authors, and might lead to issues for library
+users if their architecture is not included in the prebuilt set of binaries.
+
+### OCaml compiler version
+
+ReScript is compatible with the 4.06 version of the OCaml compiler, while
+Melange is compatible with the version 4.14 (as of May 2023).
+
+The (Melange roadmap)[todo-fix-me.md] includes a milestone to upgrade to the
+latest version of the compiler 5.0.
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:568]:
```suggestion
latest version of the compiler to the 5.x release line.
```

File: docs/melange-for-x-developers.md [hunk 9]
```diff
@@ -0,0 +1,633 @@
...
+tools](https://ocaml.org/docs/up-and-running#configuring-your-editor), which
+makes possible to work in projects that include OCaml and Melange code using the
+same editor configuration.
+
+ReScript has its [own set of editor
+plugins](https://rescript-lang.org/docs/manual/latest/editor-plugins).
+
+### Feature choice and alignment with OCaml
+
+ReScript goal is to model the language to bring it as close to JavaScript as
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:582]:
```suggestion
ReScript's goal is to model the language to bring it as close to JavaScript as
```

File: docs/melange-for-x-developers.md [hunk 10]
```diff
@@ -0,0 +1,633 @@
...
+  operators](https://v2.ocaml.org/manual/bindingops.html) (introduced in OCaml
+  4.13).
+- Optional fields in records, like `type t = { x : int, @optional y : int }`.
+- Uncurried by default.
+
+The restriction above only applies to features that compromise compatibility
+with OCaml, but otherwise Melange can incorporate bugfixes or new functionality
+from ReScript.
+
+On the other hand, as Melange goal is to keep up with the version of the OCaml
```
Comment (anmonteiro) [inline docs/melange-for-x-developers.md:608]: We aren't in a position to make this statement.

File: docs/melange-for-x-developers.md (hunk 6 above)
Comment (jchavarri) [reply docs/melange-for-x-developers.md]: https://github.com/melange-re/documentation-site/pull/39/commits/fbd10be37b0336e9d061a2774fc13146079d896c

File: docs/melange-for-x-developers.md (hunk 10 above)
Comment (jchavarri) [reply docs/melange-for-x-developers.md:608]: I agree. What do you think about https://github.com/melange-re/documentation-site/pull/39/commits/ae675c534e65567f187eb83b77ad6145735c165e?

File: docs/melange-for-x-developers.md (hunk 1 above)
Comment (jchavarri) [reply docs/melange-for-x-developers.md:298]: Not sure I understand, can you elaborate please?

File: docs/melange-for-x-developers.md [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+## For JavaScript developers
+
+Melange is a thin layer over [OCaml](https://ocaml.org/), a strongly typed
+functional programming language with an emphasis on expressiveness and safety.
+Melange’s goal is to help web developers build and maintain JavaScript
+applications safely, thanks to OCaml expressive and powerful type system.
+
+Melange allows to build applications using either OCaml syntax or Reason syntax.
+If you don’t know which one to choose, we recommend
+[Reason](https://reasonml.github.io/en/), as it has been designed with
```
[the remark this answers is not in the document]
Comment (jchavarri) [reply docs/melange-for-x-developers.md:21]: Added ref in https://github.com/melange-re/documentation-site/pull/39/commits/fac0c88f58ed1d2ec8d371b8fecda63e963986bd.

File: docs/melange-for-x-developers.md [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+
+| TypeScript                             | OCaml (Reason syntax)          |
+| -------------------------------------- | ------------------------------ |
+| `type length = <T>(_: T[]) => number;` | `let length: list('a) => int;` |
+
+## For Js_of_ocaml developers
+
+There are many similarities between Js_of_ocaml and Melange:
+
+- Both are available as libraries in the official opam repository.
```
[the remark this answers is not in the document]
Comment (jchavarri) [reply docs/melange-for-x-developers.md:478]: https://github.com/melange-re/documentation-site/pull/39/commits/fac0c88f58ed1d2ec8d371b8fecda63e963986bd

File: docs/melange-for-x-developers.md (hunk 7 above)
Comment (jchavarri) [reply docs/melange-for-x-developers.md]: reworded in https://github.com/melange-re/documentation-site/pull/39/commits/dde8ae32d450e5e5dff9c2837f6970ec7bf916d2.

File: docs/melange-for-x-developers.md (hunk 2 above)
Comment (jchavarri) [reply docs/melange-for-x-developers.md]: https://github.com/melange-re/melange-re.github.io/pull/39/commits/dde8ae32d450e5e5dff9c2837f6970ec7bf916d2

**Pull request merged**